models/classes/runner/time/TimerAdjustmentService.php

```diff
@@ -0,0 +1,119 @@
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2020 (original work) Open Assessment Technologies SA ;
+ */
+
+namespace oat\taoQtiTest\models\runner\time;
+
+use oat\oatbox\service\ConfigurableService;
+use oat\taoDelivery\model\execution\DeliveryExecutionInterface;
+use oat\taoQtiTest\models\runner\StorageManager;
+use oat\taoQtiTest\models\TestSessionService;
+use oat\taoTests\models\runner\time\TimerAdjustmentServiceInterface;
+use qtism\data\QtiIdentifiable;
+
+class TimerAdjustmentService extends ConfigurableService implements TimerAdjustmentServiceInterface
+{
+    /** @var QtiTimer $timer */
+    private $timer;
+
+    /**
+     * @inheritDoc
+     */
+    public function increase($deliveryExecution, $seconds, $source = null)
+    {
+        $this->timer = $this->getTestSessionService()->getTestSession($deliveryExecution)->getTimer();
+
+        return $this->register($deliveryExecution, AdjustmentMap::ACTION_INCREASE, $seconds);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function decrease($deliveryExecution, $seconds, $source = null)
+    {
+        $this->timer = $this->getTestSessionService()->getTestSession($deliveryExecution)->getTimer();
+
+        $maxSeconds = $this->findMaximumPossibleDecrease($deliveryExecution);
+        if ($maxSeconds < $seconds) {
+            $seconds = $maxSeconds;
+        }
+
+        return $this->register($deliveryExecution, AdjustmentMap::ACTION_DECREASE, $seconds);
+    }
+
+    /**
+     * @param DeliveryExecutionInterface $deliveryExecution
+     * @param string $action
+     * @param integer $seconds
+     * @return bool
+     */
+    private function register($deliveryExecution, $action, $seconds)
+    {
+        $testSession = $this->getTestSessionService()->getTestSession($deliveryExecution);
+        $this->putAdjustmentToTheMap($testSession->getCurrentAssessmentItemRef(), $action, $seconds);
+        $this->putAdjustmentToTheMap($testSession->getCurrentAssessmentSection(), $action, $seconds);
+        $this->putAdjustmentToTheMap($testSession->getCurrentTestPart(), $action, $seconds);
+        $this->putAdjustmentToTheMap($testSession->getAssessmentTest(), $action, $seconds);
+        $this->timer->save();
+
+        $this->getServiceLocator()->get(StorageManager::SERVICE_ID)->persist();
+
+        \common_Logger::i(var_export($this->timer->getAdjustmentMap()->toArray(), true));
+
+        return true;
```

Comment:
It always returns `true`. Should we return false if timer cannot be not adjusted or something failed?

Reply:
Right, I've updated the implementation with a case where a decrease can not be registered and `false` is returned